core/src/main/scala/kafka/log/UnifiedLog.scala

```diff
@@ -1390,7 +1424,15 @@ class UnifiedLog(@volatile var logStartOffset: Long,
   private def deleteOldSegments(predicate: (LogSegment, Option[LogSegment]) => Boolean,
                                 reason: SegmentDeletionReason): Int = {
     def shouldDelete(segment: LogSegment, nextSegmentOpt: Option[LogSegment]): Boolean = {
-      highWatermark >= nextSegmentOpt.map(_.baseOffset).getOrElse(localLog.logEndOffset) &&
+      val upperBoundOffset = nextSegmentOpt.map(_.baseOffset).getOrElse(localLog.logEndOffset)
+
+      // Check not to delete segments which are not yet copied to tiered storage
+      val isSegmentTieredToRemoteStorage =
```

Comment:
Local log size is based on the local retention configs and those are always less than or equal to the complete log retention.

I'm unclear about the rationale behind retaining data in local storage using an overall retention size where there are no remote log segments. Please provide clarification.

Reply:
Here is what I mean. Ideally, the retention behavior should be unchanged with remote storage. Consider the following case without remote storage. Let's say retentionSize is 100MB and we have only 1 segment of 90MB. The retention logic won't trigger the deletion of the last segment.

Now, consider the same situation with remote storage enabled, but no remote segments. If localRetention is 20MB, the retention logic will delete last segment of 90MB. Since the data is not in remote storage. We have deleted the data a bit earlier than expected.

A similar issue exists for time-based retention. If remote storage is enabled, but no remote segments, the time-based retention is now based on localRentionTime, not retentionTime. Since the former can be smaller than the latter, it means that we could delete the data earlier than expected.

Reply:
Good point! I think it's worth filing a bug in JIRA. WDYT @satishd ?

Reply:
Let me rephrase what you mentioned here

retention.bytes= 100MB
segment1 - 90MB

When remote storage is not enabled, then this segment is not deleted from local log segments becuas eof the retention size check. 

retention.bytes= 100MB
local.retention.bytes= 20MB
segment1 - 90MB

When remote storage is enabled, and there are no segments uploaded to remote storage. That means it will not allow this segment to be deleted as it is not yet copied to remote storage based on the introduced check in this PR. 

If it is copied to remote storage, that means it is not an active segment and there are one or more local segments after this segment. This segment will be eligible for deletion based on the local retention policy as it is already copied to remote storage earlier.

@junrao Am I missing antyhing here?

Reply:
Let me rephrase what you mentioned here

retention.bytes= 100MB
segment1 - 90MB

When remote storage is not enabled, then this segment is not deleted from local log segments becuas eof the retention size check.

retention.bytes= 100MB
local.retention.bytes= 20MB
segment1 - 90MB

When remote storage is enabled, and there are no segments uploaded to remote storage. That means it will not allow this segment to be deleted as it is not yet copied to remote storage based on the introduced check in this PR.

If it is copied to remote storage, that means it is not an active segment and there are one or more local segments after this segment. This segment will be eligible for deletion based on the local retention policy as it is already copied to remote storage earlier.

@junrao Am I missing anything here?

Reply:
@satishd : Sorry, I didn't give the right example. This is the case.

Without remote storage,
  retention.bytes= 100MB
  segment1 - 200MB
We will delete segment1 (even if it's the active segment).

With remote storage, 
  retention.bytes= 100MB
  local.retention.bytes= 20MB
  segment1 - 200MB
If segment1 is the active segment, it won't be deleted until it rolls and is uploaded to the remote store.

It's a very subtle difference.

Reply:
Thanks @junrao for the clarification. 

In the above case with remote storage enabled, it will eventually be deleted from local and remote storages, and updates log-start-offset and local-log-start-offset respectively.

Reply:
Yes, I agree that it's not a large and common issue. So, we can leave it as it is for now.